war/src/main/js/api/pluginManager.js

```diff
@@ -7,18 +7,49 @@ var jenkins = require('../util/jenkins');
 // TODO: Get plugin info (plugins + recommended plugin list) from update center.
 // For now, we statically store them in the wizard.
 
-var plugins = require('./plugins.js');
-plugins.names =[];
-for (var i = 0; i < plugins.availablePlugins.length; i++) {
-    var pluginCategory = plugins.availablePlugins[i];
-    var categoryPlugins = pluginCategory.plugins;
-    for (var ii = 0; ii < categoryPlugins.length; ii++) {
-        var pluginName = categoryPlugins[ii].name;
-        if (plugins.names.indexOf(pluginName) === -1) {
-            plugins.names.push(pluginName);
-        }
-    }
-}
+//Get plugin info (plugins + recommended plugin list) from update centers.
+var plugins;
+ 
+exports.initialPluginList = function(handler) {
+ jenkins.get('/setupWizard/platformPluginList', function(response) {
+   if(response.status !== 'ok') {
+     handler.call({ isError: true, data: response.data });
+     return;
+   }
+
+   handler.call({ isError: false }, response.data);
+ }, {
+   timeout: pluginManagerErrorTimeoutMillis,
+   error: function(xhr, textStatus, errorThrown) {
+     handler.call({ isError: true, errorMessage: errorThrown });
+   }
+ });
+};
+ 
+// Call this to initialize the plugin list
+exports.init = function(handler) {
+	exports.initialPluginList(function(initialPluginCategories) {
+		plugins = {};
+		plugins.names = [];
+		plugins.recommendedPlugins = [];
+		plugins.availablePlugins = initialPluginCategories;
+		for (var i = 0; i < initialPluginCategories.length; i++) {
```

Comment:
Yes, it's guaranteed. Even if the categories list is empty due to the custom update site (it is not going to work well), there will be empty list from what I see